src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -0,0 +1,241 @@
+    }
+
+    @Override
+    public Set<Index> getIndexes()
+    {
+        return ImmutableSet.copyOf(indices);
+    }
+
+    @Override
+    public void addIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        indices.add((StorageAttachedIndex) index);
+    }
+
+    @Override
+    public void removeIndex(Index index)
+    {
+        assert index instanceof StorageAttachedIndex;
+        boolean removed = indices.remove(index);
+        assert removed : "Cannot remove non-existing index " + index;
+        /*
+         * For the in-memory implementation we only need to unsubscribe from the tracker
+         */
+        if (indices.isEmpty())
+        {
+            baseCfs.getTracker().unsubscribe(this);
+        }
+    }
+
+    @Override
+    public void invalidate()
+    {
+        // in case of dropping table, sstable contexts should already been removed by SSTableListChangedNotification.
+        queryMetrics.release();
+        stateMetrics.release();
+        baseCfs.getTracker().unsubscribe(this);
+    }
+
+    @Override
+    public boolean containsIndex(Index index)
+    {
+        return index instanceof StorageAttachedIndex && indices.contains(index);
+    }
+
+    @Override
+    public Index.Indexer indexerFor(Predicate<Index> indexSelector,
+                                    DecoratedKey key,
+                                    RegularAndStaticColumns columns,
+                                    int nowInSec,
+                                    WriteContext ctx,
+                                    IndexTransaction.Type transactionType,
+                                    Memtable memtable)
+    {
+        final Set<Index.Indexer> indexers =
+                indices.stream().filter(indexSelector)
+                                .map(i -> i.indexerFor(key, columns, nowInSec, ctx, transactionType, memtable))
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toSet());
+
+        return indexers.isEmpty() ? null : new StorageAttachedIndex.IndexerAdapter()
+        {
+            @Override
+            public void insertRow(Row row)
+            {
+                forEach(indexer -> indexer.insertRow(row));
+            }
+
+            @Override
+            public void updateRow(Row oldRow, Row newRow)
+            {
+                forEach(indexer -> indexer.updateRow(oldRow, newRow));
+            }
+
+            private void forEach(Consumer<Index.Indexer> action)
+            {
+                indexers.forEach(action);
+            }
+        };
+    }
+
+    @Override
+    public StorageAttachedIndexQueryPlan queryPlanFor(RowFilter rowFilter)
+    {
+        return StorageAttachedIndexQueryPlan.create(baseCfs, queryMetrics, indices, rowFilter);
+    }
+
+    @Override
+    public SSTableFlushObserver getFlushObserver(Descriptor descriptor, LifecycleNewTracker tracker, TableMetadata tableMetadata)
+    {
+        return null;
```

Comment:
Just returning `null` for now, and will add back later when we have the on-disk bits, right?

Reply:
Correct. Returning `null` causes it to be ignored for now.